api/app.py

```diff
@@ -42,25 +49,65 @@
 PalaceXrayProfiler.configure(app)
 
 
-@app.before_first_request
+def initialize_application() -> Flask:
+    with app.app_context(), flask_babel.force_locale("en"):
+        initialize_database()
+        initialize_circulation_manager()
+        initialize_admin()
+    return app
+
+
 def initialize_database(autoinitialize=True):
     testing = "TESTING" in os.environ
-
     db_url = Configuration.database_url()
     if autoinitialize:
         SessionManager.initialize(db_url)
     session_factory = SessionManager.sessionmaker(db_url)
     _db = flask_scoped_session(session_factory, app)
     app._db = _db
-
     log_level = LogConfiguration.initialize(_db, testing=testing)
     debug = log_level == "DEBUG"
     app.config["DEBUG"] = debug
     app.debug = debug
     _db.commit()
+
     logging.getLogger().info("Application debug mode==%r" % app.debug)
 
 
+def initialize_admin(_db=None):
+    if getattr(app, "manager", None) is not None:
+        setup_admin_controllers(app.manager)
+    _db = _db or app._db
+    _db.execute("LOCK TABLE configurationsettings IN ACCESS EXCLUSIVE MODE;")
+    # The secret key is used for signing cookies for admin login
+    app.secret_key = ConfigurationSetting.sitewide_secret(_db, Configuration.SECRET_KEY)
+    # Create a default Local Analytics service if one does not
+    # already exist.
+
+    local_analytics = LocalAnalyticsProvider.initialize(_db)
+    _db.commit()
```

Comment:
I worry a bit about having these locks around in the code. Could we improve the situation a bit creating a context manager, pulling it back into the ./core/model module somewhere, and just invoking it from here?

So you might have some like this in `core/model/configuration.py`:

```python
from contextlib import contextmanager

if TYPE_CHECKING:
    from sqlalchemy.orm import Session

...

@contextmanager
def access_exclusive_lock_configurationsettings(session: Session):
    session.begin_nested()
    session.execute("LOCK TABLE configurationsettings IN ACCESS EXCLUSIVE MODE;")
    try:
        yield session
    finally:
        session.commit()
```


Then `initialize_admin` might look something like this (note that the `_db.commit()` at the end is gone):
```python
def initialize_admin(_db=None):
    if getattr(app, "manager", None) is not None:
        setup_admin_controllers(app.manager)

    _db = _db or app._db
    with access_exclusive_lock_configurationsettings(_db) as db_session:
        # The secret key is used for signing cookies for admin login
        app.secret_key = ConfigurationSetting.sitewide_secret(db_session, Configuration.SECRET_KEY)
        # Create a default Local Analytics service if one does not
        # already exist.
        local_analytics = LocalAnalyticsProvider.initialize(db_session)
```

Reply:
@tdilauro : that's very slick.  I'm willing it give it a try.

Reply:
https://github.com/ThePalaceProject/circulation/pull/898/commits/e2445dfc03075e19076ae597431850f7b2a0594d

Reply:
Added to the latest.